mayastor/src/nats.rs

```diff
@@ -0,0 +1,284 @@
+                Ok(client) => Some(client),
+                Err(err) => {
+                    error!("{}", err);
+                    delay_for(Duration::from_secs(self.hb_interval)).await;
+                    continue;
+                }
+            };
+        }
+        info!("Connected to the NATS server {}", self.server);
+
+        info!(
+            "Registering '{}' and grpc server {} ...",
+            self.node, self.grpc_endpoint
+        );
+        loop {
+            if let Err(err) = self.register().await {
+                error!("Registration failed: {:?}", err);
+            };
+            let _res = select! {
+                () = delay_for(Duration::from_secs(self.hb_interval)).fuse() => (),
+                msg = receiver.next() => {
+                    match msg {
+                        Some(_) => warn!("Messages have not been implemented yet"),
+                        None => {
+                            debug!("Terminating the NATS client");
+                            break;
+                        }
+                    }
+                }
+            };
+        }
+
+        if let Err(err) = self.deregister().await {
+            error!("Deregistration failed: {:?}", err);
+        };
+        Ok(())
+    }
+
+    /// Try to connect to the NATS server including DNS resolution step if
+    /// needed.
+    async fn connect(&self) -> Result<NatsClient, Error> {
+        // Resolve the hostname of the server - nats lib won't do that for us
+        let server_ip = match SocketAddr::from_str(&self.server) {
+            Ok(_) => self.server.clone(),
+            Err(_) => resolve(&self.server).await?,
+        };
+
+        debug!("Connecting to the message bus");
+        let config = NatsConfigBuilder::default()
+            .reconnection_period(Duration::from_secs(self.hb_interval))
+            .server(&server_ip)
+            .build()
+            .map_err(|cause| Error::BuildClient {
+                server: self.server.clone(),
+                cause,
+            })?;
+
+        connect(config).await.map_err(|err| Error::ConnectFailed {
+            server: self.server.clone(),
+            cause: err,
+        })
+    }
+
+    /// Send a register message to the NATS server.
+    async fn register(&mut self) -> Result<(), Error> {
+        let payload = RegisterArgs {
+            id: self.node.clone(),
+            grpc_endpoint: self.grpc_endpoint.clone(),
+        };
+        match &mut self.client {
+            Some(client) => client
+                .publish("register", serde_json::to_vec(&payload).unwrap())
+                .await
+                .map_err(|cause| Error::QueueRegister {
+                    cause,
+                })?,
+            None => return Err(Error::NotStarted {}),
+        }
+        // Note that the message was only queued and we don't know if it was
+        // really send to the NATS server (limitation of the nats lib)
+        debug!(
```

Comment:
Is this printed every HB_INTERVAL? If so should it be a trace rather than a debug?

Reply:
yes, it is printed every HB_INTERVAL that is in seconds and by default it is set to 10s. That's not very often compared to trace messages in IO path, but you are right that it is repeating information.